### ISSUE TITLE
Integrate http://github.com/decoupled/redwood-core (Language Server)

### ISSUE THREAD

**Issue by aldonline** [redwoodjs/redwood#795]
I have been working on a [Language Server](https://langserver.org/) for Redwood.js.
The code is in currently in [a different repo](https://github.com/decoupled/redwood-core) (It was refactored out from Decoupled Studio). 

Goals:

1. Primary: When a user installs the Redwood CLI, the language server should be installed too. This will allow Decoupled Studio (and eventually other tools) to reach into node_modules and start a language sever that matches the installed Redwood version (so the diagnostics you see in VSCode will be coming directly from the Redwood codebase, for example).
2. Secondary: The [core classes](https://github.com/decoupled/redwood-core/tree/master/src/core) provide a robust foundation for other parts of Redwood (ex: CLI, generators, and even the router<sup>1</sup>). Over time, this package should grow to contain most features that are now scattered. In a way, redwood-core is a more ambitious take on redwood-internal, and it should subsume it over time. (for now, redwood-core uses some of the methods in redwood internal, but it re-implements most of them).

Plan:

We need to do some pair programming (@aldonline + @peterp) to make the following happen:

- [x] Move the code from the decoupled github org to redwoodjs/redwood (Create a new package called `project-model`, copy paste the code, and add it to the build process so it gets published to NPM) (1-2hrs pair programming) ff92117
- [x] Make redwood-core a direct dependency of redwood-cli, so it gets installed in node_modules. (10 minutes, anyone) #819
- [ ] Make Decoupled Studio detect and use the now built-in language server that ships with Redwood. At this point, users with Decoupled Studio will start seeing all kinds of cool diagnostics, quick-fixes, and an improved project outline (Goal 1). More importantly, these features can be improved from release to release by the Redwood team directly. (4 hours @aldonline)
- [ ] Use the [project-model/core](https://github.com/decoupled/redwood-core/tree/master/src/core) classes to provide diagnostic checking for some of the CLI lifecycle methods (ex: rw build). This is the lowest hanging fruit, and a good first step towards Goal 2. (1 hour, pair programming @aldonline + anyone from the redwood core team)
4. Add more diagnostics, more quick fixes, more documentation to to the core classes, etc. There will be a very clear incentive once users start seeing features immediately pop up in VSCode.

---

<sup>1</sup>. The core classes are the equivalent of an AST. It models the relationship between the different concepts and their restrictions. As such, it provides a good foundation for everything else. For example, generators can use the project model to query information about existing pages and models, the CLI can use it to check that there are no conflicting Routes (at compile time), etc.

**Comment by peterp** [comment]
Hey @aldonline,

I've slightly modified your suggested approach and done the following.

- [x] Created a new package called `@redwoodjs/project-model` #ff92117b4804c11593d9efd8258c23d57d4c49f8
- [ ] Install in `@redwoodjs/cli`

**Comment by aldonline** [comment]
Hi Peter!
Thanks. That's a lot of work :)
I believe I should have emphasized the "pair programing" part a bit more. My bad. Let me elaborate:
I am working on a PR with this code and instead of just sending it your way I wanted to work on it together (hence the first item) to iron out some details (like code style), integrate the build, go over the code to make sure we were aligned, and talk about the workflow moving forward (the most important part). When I said copy paste I meant doing it into the PR. (As opposed to moving the actual decoupled/redwood-core repo to the organization).
I'm of two minds right now. On one hand I just want to move forward, but on the other I think we missed the opportunity to do the integration together and on a technical level we lost some git history. This huge part of the codebase just came out of thin air, and when git-blaming they will only find you. It is not an ego thing (as you can see I don't like to appear as contributor to projects because I don't like to take on extra work maintaining the code later, so I usually encourage people to just take it over). But in this case I think we have a long way to go still. I intend to keep working on this.
So, what do we do?

**Comment by aldonline** [comment]
Actually, nevermind. I just jumped to respond quickly as soon as I saw this but after looking at the code I think 1) you did a much deeper integration that I could have done (which is awesome, that's exactly what was holding me back from sending the PR) and 2) you added my GH handle to the commit title, and the issue is linked, so I think history is preserved.
So, we're good!
We should still sync up via chat to go over the workflow. Most of this package is stable (specially the ./core) but the outline, for example, will have some major changes as I make it work with studio.
Thanks for your work and for embracing the contribution and ideas. I'm really looking forward to seeing this evolve within the codebase :)
